src/ui_manager.h

```diff
@@ -32,39 +88,75 @@ class ui_adaptor
-        //
-        // The redraw callback should also not call `position_from_window`,
-        // otherwise it may cause UI glitch if the window position changes.
+        /**
+         * Set redraw and resize callbacks. The resize callback should
+         * call `position` or `position_from_window` to set the size of the UI,
+         * and (re-)calculate any UI data that is related to the screen size,
+         * including `catacurses::window` instances. In most cases, you should
+         * also call `mark_resize` along with `on_screen_resize` so the UI is
+         * initialized by the resizing callback when redrawn for the first time.
+         *
+         * The redraw callback should only redraw to the area specified by the
+         * `position` or `position_from_window` call. Content drawn outside this
+         * area may not render correctly when resizing or exiting from other UIs.
+         * Transparency is not currently supported.
+         *
+         * These callbacks should NOT:
+         * - Construct new `ui_adaptor` instances
+         * - Deconstruct old `ui_adaptor` instances
+         * - Call `redraw` or `screen_resized`
+         * - (Redraw callback) call `position_from_window`
+         * - Call any function that does these things, except for `debugmsg`
+         *
+         * Otherwise, display glitches or even crashes might happen.
+         *
+         * Calling `debugmsg` inside the callbacks is (semi-)supported, but may
+         * cause display glitches after the debug message is closed.
+         **/
         void on_redraw( const redraw_callback_t &fun );
+        /* See `on_redraw`. */
         void on_screen_resize( const screen_resize_callback_t &fun );
 
-        // Mark this ui_adaptor for resizing the next time `redraw()` is called.
-        // This is useful for deferring initialization of the UI when explicit
-        // initialization is not possible or wanted.
+        /**
+         * Mark this `ui_adaptor` for resizing the next time it is redrawn.
+         * This is normally called alongside `on_screen_resize` to initialize
+         * the UI on the first redraw. You should also use this function to
+         * explicitly request a reinitialization if any value the screen resize
+         * callback depends on (apart from the screen size) has changed.
+         **/
         void mark_resize() const;
 
-        // Invalidate this UI so it gets redrawn the next time screen is refreshed,
-        // unless an upper UI completely occludes this UI. May also cause upper UIs
-        // to redraw.
+        /**
+         * Invalidate this UI so it gets redrawn the next redraw unless an upper
+         * UI completely occludes this UI. May also cause upper UIs to redraw.
+         * Can be used to mark lower UIs for redrawing when their associated data
+         * has changed.
+         **/
         void invalidate_ui() const;
 
-        // Reset all callbacks and dimensions
+        /** 
```

Comment:
Trailing space made astyle angry.
```suggestion
        /**
```